### PR TITLE
[Issue #3474] Fix illegal message attack

### PR DESCRIPTION
**Pull request by Aaron-He**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

fix illegal message attack

## Brief changelog

Check when decoding the protocol 

## Verifying this change

Yes

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java [hunk 1]
```diff
@@ -149,6 +150,9 @@ public static RemotingCommand rocketMQProtocolDecode(final byte[] headerArray) {
         // String remark
         int remarkLength = headerBuffer.getInt();
         if (remarkLength > 0) {
+            if (remarkLength > headerArray.length) {
```
Comment (Git-Yang) [inline remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java:153]: Will it cause the similar problem if you modify other values in the header?

Comment (Aaron-He) [reply remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java:153]: > Will it cause the similar problem if you modify other values in the header?

In addition to remarkLength and extFieldsLength, modifying other values in the header will not create large objects, but will only fail to read and throw exceptions

File: remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java [hunk 2]
```diff
@@ -157,6 +162,9 @@ public static RemotingCommand rocketMQProtocolDecode(final byte[] headerArray) {
         // HashMap<String, String> extFields
         int extFieldsLength = headerBuffer.getInt();
         if (extFieldsLength > 0) {
+            if (extFieldsLength > headerArray.length) {
```
Comment (Git-Yang) [inline remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java:165]: Would it be more appropriate to replace `headerArray.length` with `headerBuffer.capacity() - headerBuffer.position()`?

File: acl/src/test/java/org/apache/rocketmq/acl/plain/PlainAccessValidatorTest.java [hunk 3]
```diff
@@ -17,19 +17,7 @@
...
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
-
-import org.apache.rocketmq.acl.common.AclClientRPCHook;
-import org.apache.rocketmq.acl.common.AclConstants;
-import org.apache.rocketmq.acl.common.AclException;
-import org.apache.rocketmq.acl.common.AclUtils;
-import org.apache.rocketmq.acl.common.SessionCredentials;
+import org.apache.rocketmq.acl.common.*;
```
Comment (duhenglucky) [inline acl/src/test/java/org/apache/rocketmq/acl/plain/PlainAccessValidatorTest.java:20]: Aggregating all imports is not a good practice

File: acl/src/test/java/org/apache/rocketmq/acl/plain/PlainAccessValidatorTest.java [hunk 4]
```diff
@@ -39,16 +27,21 @@
...
 import org.apache.rocketmq.common.protocol.heartbeat.ProducerData;
 import org.apache.rocketmq.common.protocol.heartbeat.SubscriptionData;
+import org.apache.rocketmq.remoting.exception.RemotingCommandException;
 import org.apache.rocketmq.remoting.protocol.RemotingCommand;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 
+import java.nio.ByteBuffer;
+import java.util.*;
```
Comment (duhenglucky) [inline acl/src/test/java/org/apache/rocketmq/acl/plain/PlainAccessValidatorTest.java:37]: Same with last comment

File: remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java (hunk 2 above)
Comment (Aaron-He) [reply remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java:165]: > Would it be more appropriate to replace `headerArray.length` with `headerBuffer.capacity() - headerBuffer.position()`?

`headerBuffer.capacity() - headerBuffer.position()` is more precise than `headerArray.length`.
Howerver, `headerBuffer.capacity() - headerBuffer.position()` has more subtraction operation than `headerArray.length`. Considering performance, `headerArray.length` may be more appropriate.

File: acl/src/test/java/org/apache/rocketmq/acl/plain/PlainAccessValidatorTest.java (hunk 4 above)
Comment (Aaron-He) [reply acl/src/test/java/org/apache/rocketmq/acl/plain/PlainAccessValidatorTest.java:37]: > Same with last comment

headerBuffer.capacity() - headerBuffer.position() is more precise than headerArray.length.
Howerver, headerBuffer.capacity() - headerBuffer.position() has more subtraction operation than headerArray.length. Considering performance, headerArray.length may be more appropriate.

File: acl/src/test/java/org/apache/rocketmq/acl/plain/PlainAccessValidatorTest.java (hunk 3 above)
Comment (Aaron-He) [reply acl/src/test/java/org/apache/rocketmq/acl/plain/PlainAccessValidatorTest.java:20]: Yes, I'll push it again.

**Review by guyinyou — APPROVED** [review]

**Comment by dongeforever** [comment]
LGTM.
All the size delimited serialization has such vulnerability

File: remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RemotingCommand.java [hunk 5]
```diff
@@ -136,12 +137,12 @@ public static RemotingCommand createResponseCommand(int code, String remark) {
...
+    public static RemotingCommand decode(final byte[] array) throws RemotingCommandException {
         ByteBuffer byteBuffer = ByteBuffer.wrap(array);
         return decode(byteBuffer);
     }
 
-    public static RemotingCommand decode(final ByteBuffer byteBuffer) {
+    public static RemotingCommand decode(final ByteBuffer byteBuffer) throws RemotingCommandException {
         int length = byteBuffer.limit();
         int oriHeaderLen = byteBuffer.getInt();
         int headerLength = getHeaderLength(oriHeaderLen);
```
Comment (dongeforever) [inline remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RemotingCommand.java:148]: The headerLength need an overflow check too.

if (headerLength >= length) {
   throw xx
}

**Review by dongeforever — APPROVED** [review]
The header length need a check too.

It could be done in this pr or later pr.

**Comment by zhouxinyu** [comment]
Great catch. But this PR includes too many code style modifications, do you import the rmq_codeStyle.xml to your IDE?

**Review by duhenglucky — APPROVED** [review]

**Comment by duhenglucky** [comment]
> Great catch. But this PR includes too many code style modifications, do you import the rmq_codeStyle.xml to your IDE?

Merge it first for security reason, code styles will be fix in another PR :)

**Pull request merged**